File: mobile-ios/Services/RealtimeStreamService.swift
import Foundation
import Combine

@MainActor
class RealtimeStreamService: ObservableObject {
    static let shared = RealtimeStreamService()

    @Published var isConnected: Bool = false
    @Published var events: [TelemetryEvent] = []
    @Published var unreadCount: Int = 0

    private var webSocketTask: URLSessionWebSocketTask?
    private var simulationTimer: Timer?
    private let serverURL = "ws://localhost:5050/hubs/notifications"

    init() {
        startConnection()
    }

    func startConnection() {
        guard let url = URL(string: serverURL) else {
            startSimulationStream()
            return
        }

        let session = URLSession(configuration: .default)
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()
        isConnected = true

        receiveMessage()
        startSimulationStream()
    }

    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            Task { @MainActor in
                guard let self = self else { return }
                switch result {
                case .success(let message):
                    switch message {
                    case .string(let text):
                        if let data = text.data(using: .utf8),
                           let event = try? JSONDecoder().decode(TelemetryEvent.self, from: data) {
                            self.addEvent(event)
                        }
                    case .data(let data):
                        if let event = try? JSONDecoder().decode(TelemetryEvent.self, from: data) {
                            self.addEvent(event)
                        }
                    @unknown default:
                        break
                    }
                    self.receiveMessage()
                case .failure:
                    self.isConnected = false
                }
            }
        }
    }

    func addEvent(_ event: TelemetryEvent) {
        events.insert(event, at: 0)
        if events.count > 50 {
            events.removeLast()
        }
        unreadCount += 1
    }

    func clearUnreadCount() {
        unreadCount = 0
    }

    private func startSimulationStream() {
        simulationTimer?.invalidate()
        simulationTimer = Timer.scheduledTimer(withTimeInterval: 4.5, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self else { return }
                let sampleEvents: [(String, String, EventSeverity, String)] = [
                    ("FastAPI-AI", "IsolationForest", .info, "Anomaly model evaluated 128 telemetry vectors in 1.4ms."),
                    ("SignalR-Hub", "Notification", .info, "Broadcasted system health metric (99.4%) to 4 active clients."),
                    ("PGVector", "HNSW-Index", .info, "Cosine similarity search executed across 1536-dimensional embeddings."),
                    ("DotNet-Api", "WorkflowExecution", .warning, "Pipeline 'Document Sentiment' retry attempt #1 triggered."),
                    ("FastAPI-AI", "IsolationForest", .critical, "Telemetry threshold anomaly detected on node-us-east-2! Anomaly Score: 0.892.")
                ]

                let sample = sampleEvents.randomElement()!
                let formatter = DateFormatter()
                formatter.dateFormat = "HH:mm:ss"
                let timestamp = formatter.string(from: Date())

                let newEvent = TelemetryEvent(
                    id: UUID().uuidString,
                    timestamp: timestamp,
                    sourceComponent: sample.0,
                    category: sample.1,
                    severity: sample.2,
                    message: sample.3,
                    executionTimeMs: Double.random(in: 0.5...4.2)
                )
                self.addEvent(newEvent)
            }
        }
    }

    deinit {
        simulationTimer?.invalidate()
        webSocketTask?.cancel(with: .goingAway, reason: nil)
    }
}

File: mobile-android/app/src/main/java/com/nexusops/data/RealtimeStreamService.kt
package com.nexusops.data

import com.nexusops.model.EventSeverity
import com.nexusops.model.TelemetryEvent
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.text.SimpleDateFormat
import java.util.*
import kotlin.random.Random

object RealtimeStreamService {
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    private val _isConnected = MutableStateFlow(true)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _events = MutableStateFlow<List<TelemetryEvent>>(emptyList())
    val events: StateFlow<List<TelemetryEvent>> = _events.asStateFlow()

    private val _unreadCount = MutableStateFlow(0)
    val unreadCount: StateFlow<Int> = _unreadCount.asStateFlow()

    private var simulationJob: Job? = null

    init {
        startSimulationStream()
    }

    private fun startSimulationStream() {
        simulationJob?.cancel()
        simulationJob = scope.launch {
            val sampleEvents = listOf(
                Quadruple("FastAPI-AI", "IsolationForest", EventSeverity.INFO, "Anomaly model evaluated 128 telemetry vectors in 1.4ms."),
                Quadruple("SignalR-Hub", "Notification", EventSeverity.INFO, "Broadcasted system health metric (99.4%) to 4 active clients."),
                Quadruple("PGVector", "HNSW-Index", EventSeverity.INFO, "Cosine similarity search executed across 1536-dimensional embeddings."),
                Quadruple("DotNet-Api", "WorkflowExecution", EventSeverity.WARNING, "Pipeline 'Document Sentiment' retry attempt #1 triggered."),
                Quadruple("FastAPI-AI", "IsolationForest", EventSeverity.CRITICAL, "Telemetry threshold anomaly detected on node-us-east-2! Anomaly Score: 0.892.")
            )

            val sdf = SimpleDateFormat("HH:mm:ss", Locale.getDefault())

            while (isActive) {
                delay(4500)
                val sample = sampleEvents.random()
                val timestamp = sdf.format(Date())

                val newEvent = TelemetryEvent(
                    id = UUID.randomUUID().toString(),
                    timestamp = timestamp,
                    sourceComponent = sample.first,
                    category = sample.second,
                    severity = sample.third,
                    message = sample.fourth,
                    executionTimeMs = Random.nextDouble(0.5, 4.2)
                )

                val currentList = _events.value.toMutableList()
                currentList.add(0, newEvent)
                if (currentList.size > 50) {
                    currentList.removeAt(currentList.size - 1)
                }

                _events.value = currentList
                _unreadCount.value += 1
            }
        }
    }

    fun clearUnreadCount() {
        _unreadCount.value = 0
    }

    private data class Quadruple<A, B, C, D>(val first: A, val second: B, val third: C, val fourth: D)
}
